File: banchus-packet/src/main/java/pe/nanamochi/banchus/components/Mods.java
package pe.nanamochi.banchus.components;

import java.util.*;
import java.util.stream.Collectors;
import lombok.AllArgsConstructor;
import lombok.Getter;

@Getter
@AllArgsConstructor
public enum Mods {
  NO_MOD(0),
  NO_FAIL(1),
  EASY(1 << 1),
  TOUCHSCREEN(1 << 2),
  HIDDEN(1 << 3),
  HARD_ROCK(1 << 4),
  SUDDEN_DEATH(1 << 5),
  DOUBLE_TIME(1 << 6),
  RELAX(1 << 7),
  HALF_TIME(1 << 8),
  NIGHTCORE(1 << 9),
  FLASHLIGHT(1 << 10),
  AUTOPLAY(1 << 11),
  SPUN_OUT(1 << 12),
  AUTOPILOT(1 << 13),
  PERFECT(1 << 14),
  KEY4(1 << 15),
  KEY5(1 << 16),
  KEY6(1 << 17),
  KEY7(1 << 18),
  KEY8(1 << 19),
  FADE_IN(1 << 20),
  RANDOM(1 << 21),
  CINEMA(1 << 22),
  TARGET(1 << 23),
  KEY9(1 << 24),
  KEY_COOP(1 << 25),
  KEY1(1 << 26),
  KEY3(1 << 27),
  KEY2(1 << 28),
  SCORE_V2(1 << 29),
  MIRROR(1 << 30);

  private final int value;

  public static List<Mods> fromBitmask(int bit) {
    if (bit == 0) return new ArrayList<>(List.of(NO_MOD));

    return Arrays.stream(values())
        .filter(mod -> mod != NO_MOD && (bit & mod.value) == mod.value)
        .collect(Collectors.toList());
  }

  public static int toBitmask(List<Mods> mods) {
    if (mods == null) return 0;
    return mods.stream().mapToInt(Mods::getValue).reduce(0, (a, b) -> a | b);
  }
}

File: banchus-packet/src/main/kotlin/pe/nanamochi/banchus/components/Mods.kt
package pe.nanamochi.banchus.components

enum class Mods(val value: UInt) {
    NO_MOD(0u),
    NO_FAIL(1u),
    EASY(1u shl 1),
    TOUCHSCREEN(1u shl 2),
    HIDDEN(1u shl 3),
    HARD_ROCK(1u shl 4),
    SUDDEN_DEATH(1u shl 5),
    DOUBLE_TIME(1u shl 6),
    RELAX(1u shl 7),
    HALF_TIME(1u shl 8),
    NIGHTCORE(1u shl 9),
    FLASHLIGHT(1u shl 10),
    AUTOPLAY(1u shl 11),
    SPUN_OUT(1u shl 12),
    AUTOPILOT(1u shl 13),
    PERFECT(1u shl 14),
    KEY4(1u shl 15),
    KEY5(1u shl 16),
    KEY6(1u shl 17),
    KEY7(1u shl 18),
    KEY8(1u shl 19),
    FADE_IN(1u shl 20),
    RANDOM(1u shl 21),
    CINEMA(1u shl 22),
    TARGET(1u shl 23),
    KEY9(1u shl 24),
    KEY_COOP(1u shl 25),
    KEY1(1u shl 26),
    KEY3(1u shl 27),
    KEY2(1u shl 28),
    SCORE_V2(1u shl 29),
    MIRROR(1u shl 30);

    companion object {
        fun fromBitmask(bitmask: UInt): List<Mods> {
            if (bitmask == 0u) return listOf(NO_MOD)

            val selectedMods =
                entries.filter { mod -> mod != NO_MOD && (bitmask and mod.value) == mod.value }

            return selectedMods.ifEmpty { listOf(NO_MOD) }
        }

        fun toBitmask(mods: List<Mods>?): UInt {
            if (mods.isNullOrEmpty()) return NO_MOD.value
            return mods.fold(0u) { acc, mod -> acc or mod.value }
        }
    }
}

fun UInt.hasAny(mask: UInt): Boolean {
    return (this and mask) != 0u
}
